aspnetcore/tutorials/first-web-api.md

```diff
@@ -1,95 +1,491 @@
+
+# [Visual Studio for Mac](#tab/visual-studio-mac)
+
+Add a `TodoContext` class in the *Models* folder:
+
+---
 
 Replace the class with the following code:
 
-[!code-csharp[](first-web-api/samples/2.0/TodoApi/Models/TodoContext.cs)]
+[!code-csharp[](first-web-api/samples/2.2/TodoApi/Models/TodoContext.cs)]
+
+## Register the database context
 
-[!INCLUDE[Register the database context](../includes/webApi/register_dbContext.md)]
+In this step, the database context is registered with the [dependency injection](xref:fundamentals/dependency-injection) container. Services (such as the DB context) that are registered with the dependency injection (DI) container are available to the controllers.
+
+Register the DB context with the service container using the built-in support for [dependency injection](xref:fundamentals/dependency-injection). Update *Startup.cs* with the following highlighted code:
+
+::: moniker range="= aspnetcore-2.2"
+[!code-csharp[](first-web-api/samples/2.2/TodoApi/Startup1.cs?highlight=5,8,25-26&name=snippet_all)]
+::: moniker-end
+::: moniker range="= aspnetcore-2.1"
+[!code-csharp[](first-web-api/samples/2.1/TodoApi/Startup.cs?highlight=7-8,25-26)]
+
+::: moniker-end
+
+The preceding code:
+
+* Removes unused `using` declarations.
+* Specifies an in-memory database is injected into the service container.
 
 ### Add a controller
 
-In Solution Explorer, right-click the *Controllers* folder. Select **Add** > **New Item**. In the **Add New Item** dialog, select the **API Controller Class** template. Name the class *TodoController*, and click **Add**.
+# [Visual Studio](#tab/visual-studio)
+
+* Right-click the *Controllers* folder.
+* Select **Add** > **New Item**.
+* In the **Add New Item** dialog, select the **API Controller Class** template.
+* Name the class *TodoController*, and click **Add**.
 
 ![Add new Item dialog with controller in search box and web API controller selected](first-web-api/_static/new_controller.png)
 
-Replace the class with the following code:
+# [Visual Studio Code](#tab/visual-studio-code)
 
-[!INCLUDE[code and get todo items](../includes/webApi/getTodoItems.md)]
+In the *Controllers* folder, create a class named `TodoController`.
 
-### Launch the app
+# [Visual Studio for Mac](#tab/visual-studio-mac)
+
+In Solution Explorer, in the *Controllers* folder, add the class `TodoController`.
+
+---
+
+Replace the `TodoController` class with the following code:
+
+[!code-csharp[](first-web-api/samples/2.2/TodoApi/Controllers/TodoController2.cs?name=snippet_todo1)]
+
+The preceding code:
+
+* Defines an API controller class without methods.
+* Creates a new Todo item when `TodoItems` is empty. You won't be able to delete all the Todo items because the constructor creates a new one if `TodoItems` is empty.
+
+In the next sections, methods are added to implement the API. The class is annotated with an [`[ApiController]`](/dotnet/api/microsoft.aspnetcore.mvc.apicontrollerattribute) attribute to:
```

Comment:
Remove code-fencing, since this is a link.

Reply:
I'd like to keep in the inner `[]` and `[\[` doesn't work. We can probably make an exception for attributes.

Reply:
Do you really need brackets when you're explicitly identifying it as an attribute?